### DISCUSSION TITLE
关于在Mac，Linux(虚拟机)，树莓派，Quark上的调试过程小结

### DISCUSSION THREAD

**Discussion by coolflyreg** [peng-zhihui/ElectronBot#169] [General]
### 调试过程小结
1、Linux在USB连接后，需要调用libusb_detach_kernel_driver，去detach掉linux内核cdc_acm驱动。至于第二个参数interface_number，0或1都可以。如果不detach kernel driver，claim的时候，会返回设备busy(-16)的错误
Mac不需要detach kernel driver
<img width="519" alt="int detach kd w  L1busb detach kernel_ driver (devHandle-handle, 0)" src="https://user-images.githubusercontent.com/4011564/165226047-b7fe3887-2ce2-4ef2-bde7-a99be8d91017.png">
detach前
<img width="608" alt="1Port 1" src="https://user-images.githubusercontent.com/4011564/165226053-04e892d0-6c96-496a-9ec0-ae6c55f7050f.png">
detach后
<img width="609" alt="Bus 03  Port 1 Dev" src="https://user-images.githubusercontent.com/4011564/165226061-f09e304e-0ddc-4de1-8e96-d23181acbd07.png">

2、Mac下需要调用usb_claim_interface才能通讯，不需要detach kernel driver。目前在打开后USB立刻调用了一次。
<img width="437" alt="int claimResult =•usb claim interface (devHand le,  0)" src="https://user-images.githubusercontent.com/4011564/165226067-425af130-d514-4c24-bad9-2e10a94ab0eb.png">

3、关于usb_claim_interface的调用参数
对于Mac和树莓派（kernel 5.15.32-v7l+），claim是0或是1都没问题，视频流畅。
对于quark（ubuntu 16.x kernel 4.14.111），claim 1的话，会奇慢无比，需要使用claim 0，然后略微有点慢，比如30帧降到20帧的感觉，但是看不到三段刷新。
对于Linux虚拟机（UOS 20 kernel 5.10.0)，不管是0还是1，都很慢，但是比quark的奇慢无比要快一点，大约1秒3帧左右，能明显看到分三段刷新。
仔细测试了下：虚拟机的Linux的和虚拟机的Win10，串口都比较慢，虚拟机的win10比虚拟机linux略快一点

4、在quark（ubuntu 16.x kernel 4.14.111）和Linux虚拟机（UOS 20 kernel 5.10.0)上连接的时候，需要先启动上位机，然后再插设备，才能正常通讯。正常通讯以后，不拔插设备，等Sample播完，可以再次运行Sample。使用如下图方法来等待连接
在Mac和树莓派（kernel 5.15.32-v7l+）上，先插后插都可以
<img width="483" alt="while (Irobot Connect())" src="https://user-images.githubusercontent.com/4011564/165226073-263287d5-56a5-4c38-a198-7732dfef495d.png">

5、如果想让Linux插ElectronBot后，不自动attach kernel driver，那么就要设置设备描述符DeviceClass为0xff，即厂商自定义。
DeviceClass改为0xff后，Win10上一样可以

### 目前不解的问题
1、[已解]Quark上的速度播放掉帧率，是不是因为H3本身的性能问题。
测试结果：没有修改代码，仅仅将视频转为了240x240的大小，再次测试，效果也差不多。结论：还是性能不足导致卡顿

**Comment by peng-zhihui** [comment]
不错，你可以把修改的描述符提个pr。
因为涉及使用OpenCV对图像做预处理，所以夸克的性能可能是不太够用，可以尝试预先用格式工厂等软件把视频分辨率降低到240x240，然后把SDK代码里面加一个判断如果分辨率是合适的就不进行resize，可以解决这个问题。

**Reply by coolflyreg** [reply]
对哦，大佬提醒的是。恍然大悟，我Quark上的视频尺寸是544 × 720，就算用mplayer播放，尺寸不一致，要进行大小转换的，也会比较卡。光想到是USB传输速率了。我再好好测试测试

**Reply by GreenShadeZhang** [reply]
> 对哦，大佬提醒的是。恍然大悟，我Quark上的视频尺寸是544 × 720，就算用mplayer播放，尺寸不一致，要进行大小转换的，也会比较卡。光想到是USB传输速率了。我再好好测试测试

请问下，我遇到一个问题，我是用树莓派zero 2w 直接接收的字节数据到队列排队写入电子，但是也是很感觉很卡顿掉帧，这和树莓派zero 2w 的性能关系大呢，还是和otg连接usb的速率低有关呢？感觉只是接收转发应该不需要太多的性能吧。
